### ISSUE TITLE
Clash between the local server's domain and s.whatsapp.net

### ISSUE THREAD

**Issue by BlackLight** [mautrix/whatsapp#492]
I have installed mautrix-whatsapp on Manjaro through the AUR package.

The bot comes online, I have logged in, I can send commands like `sync groups/contacts`, but then I get a lot of 400/403 errors from the bridge upon synchronization. Some extracts:

```
[Matrix/@***********:matrix.mydomain.com/DEBUG] req #6: POST http://localhost:8108/_matrix/client/r0/register?user_id=%40***********%3Amatrix.mydomain.com {"username":"***********","inhibit_login":true,"type":"m.login.application_service"}

[Puppet/***********@s.whatsapp.net/ERROR] Failed to ensure registered: failed to ensure registered: failed to POST /_matrix/client/r0/register: M_EXCLUSIVE (HTTP 400): Invalid user localpart for this application service.

[Puppet/***********@s.whatsapp.net/WARN] Failed to reupload avatar: failed to upload avatar to Matrix: failed to POST /_matrix/media/r0/upload: M_FORBIDDEN (HTTP 403): Application service cannot masquerade as this user (@********:matrix.platypush.tech).
```

And this actually seems to make sense: my Matrix server's domain is something like `matrix.domain.com`. The original `/register` request correctly reports the users' local domain, but then the `Puppet` replaces the local domain with `s.whatsapp.net`.

As a result, I cannot import any users and groups into my Matrix instance.

I'm not sure if I missed something in the configuration (other bridges work fine though), but `s.whatsapp.net` being propagated to the Matrix server as users' localpart seems to be an issue on the bridge side.

The errors seem to closely match those reported on https://github.com/mautrix/whatsapp/issues/6 a while ago, but no solution was proposed back then.

**Comment by tulir** [comment]
That's a config issue, probably either in the registration file `namespaces`, or the bridge's `username_template` or `homeserver` -> `domain`.

**Comment by BlackLight** [comment]
Thanks @tulir. After a bit of investigation I found out that the issue was due to `username_template`:

```yaml
bridge:
    username_template: "{{.}}[wa]"
```

But then the generated `registration.yaml` would contain:

```yaml
namespaces:
    users:
        - regex: ^@[0-9]+[wa]:matrix.platypush.tech$
          exclusive: true
```

The `[wa]` suffix is treated like a regex, and hence the namespace mismatch.

Once escaped username suffix in the `regex` as `\[wa\]` it went fine.

However, given how common it is to use templates that may contain regex-sensitive characters (parentheses, brackets, dots, question marks etc.) is it maybe a good idea to regex-sanitize the template before generating the `registration.yaml`?
